zerver/views/message_fetch.py

```diff
@@ -274,6 +279,13 @@ def by_streams(self, query: Query, operand: str, maybe_negate: ConditionTransfor
                 type_id__in=public_streams_queryset).values_list('id', flat=True).order_by('id')
             cond = column("recipient_id").in_(recipient_ids)
             return query.where(maybe_negate(cond))
+        if operand == 'web-public':
+            web_public_streams = Stream.objects.filter(realm=self.user_realm, deactivated=False, is_web_public=True)
+            recipient_ids = Recipient.objects.filter(
+                type=Recipient.STREAM,
+                type_id__in=web_public_streams).values_list('id', flat=True).order_by('id')
```

Comment:
Do this as a prep commit in the previous block and then copy that logic here, `Stream.objects.filter(...).values_list("recipient_id", flat=True)` would do what we want here without an extra query or extra code, now that we're denormalized that field.